### PR TITLE
add MADDPG algorithm

### PR DESCRIPTION
**Pull request by peterchen96**
PR Checklist

- [ ] Update NEWS.md?

The description of the implementation is in discussion #404. 

Here `MADDPG` raises an `unknown word` error... How can I fix it? @findmyway

### CODE REVIEW

**Comment by findmyway** [comment]
You can simply add those words after https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/497376243052afe715babc1a5845a1eef412b6f4/.cspell/cspell.json#L123

cc @pilgrimygy

**Comment by peterchen96** [comment]
Thanks! And ask for suggestions about the implementation and code mistakes/style.

**Comment by peterchen96** [comment]
Thanks! And ask for suggestions about the implementation and code mistakes/style.

**Comment by findmyway** [comment]
I'll review it later tonight 😃

File: src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        push!(agent.trajectory[:action], rand(action_space(env)))
+        if haskey(agent.trajectory, :legal_actions_mask)
+            lasm = legal_action_space_mask(env, player)
+            push!(agent.trajectory[:legal_actions_mask], lasm)
+        end
+    end
+
+    # update policy
+    update!(π)
+end
```
Comment (findmyway) [inline src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl:78]: How about dispatching to the inner agent's corresponding methods?

Like calling `agent(stage, env, action)` in the `for` loop.

File: src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    π.update_step % π.update_freq == 0 || return
+
+    for (_, agent) in π.agents
+        length(agent.trajectory) > agent.policy.update_after || return
+        length(agent.trajectory) > π.batch_size || return
+    end
+    
+    # get training data
+    temp_player = rand(keys(π.agents))
+    t = π.agents[temp_player].trajectory
```
Comment (findmyway) [inline src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl:92]: Simply use the first agent?

File: src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    inds = rand(π.rng, 1:length(t), π.batch_size)
+    batches = Dict((player, RLCore.fetch!(BatchSampler{SARTS}(π.batch_size), agent.trajectory, inds))
```
Comment (findmyway) [inline src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl:94]: The hardcoded `SARTS` will make the algorithm work only on environments of `MINIMAL_ACTION_SET`.

File: src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    temp_player = rand(keys(π.agents))
+    t = π.agents[temp_player].trajectory
+    inds = rand(π.rng, 1:length(t), π.batch_size)
+    batches = Dict((player, RLCore.fetch!(BatchSampler{SARTS}(π.batch_size), agent.trajectory, inds)) 
+                for (player, agent) in π.agents)
+    
+    # get s, a, s′ for critic
+    s = vcat((batches[player][1] for (player, _) in π.agents)...)
+    a = vcat((batches[player][2] for (player, _) in π.agents)...)
+    s′ = vcat((batches[player][5] for (player, _) in π.agents)...)
```
Comment (findmyway) [inline src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl:100]: `vcat` is not very efficient here. Try `Flux.batch`?

File: src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        update!(A, gs2)
+
+        # polyak averaging
+        for (dest, src) in zip(Flux.params([Aₜ, Cₜ]), Flux.params([A, C]))
+            dest .= ρ .* dest .+ (1 - ρ) .* src
+        end
+
+        s, a, s′ = send_to_host((s, a, s′))
+        mu_actions = send_to_host(mu_actions)
+        new_actions = send_to_host(new_actions)
```
Comment (findmyway) [inline src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl:169]: Are they required here?

File: src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl (hunk 1 above)
Comment (findmyway) [reply src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl:78]: Can you take a look at the `NamedPolicy` and see whether we can reuse existing code as much as possible? See also the `MultiAgentManager`

**Review by findmyway — APPROVED** [review]
Looks fine to me

File: src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl (hunk 5 above)
Comment (peterchen96) [reply src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl:169]: Thanks for your kind reviews! I'll check and update my codes later today.

**Comment by peterchen96** [comment]
Here is still a simple version of `MADDPG,` which only supports the envs of `MINIMAL_ACTION_SET.`

**Pull request merged**